### PR TITLE
Give Bodhi central configuration validation.

### LINKED ISSUE #1489: config.py should do more for us
Currently, Bodhi has config processing code scattered all over the place. [For example](https://github.com/fedora-infra/bodhi/blob/2.6.0/bodhi/server/notifications.py#L33),

```
bodhi.server.config.config.get('fedmsg_enabled')
```

This is problematic for a few reasons:

1. The code that wants to use the config is always given a string of what it is set to in the config file, regardless of what type the config makes sense to be. In the above example, it makes sense for the config to be a boolean, and it being a string and not interpreted as a bool caused #1488.
2. There is no central place in code to see what all of Bodhi's defaults are, as the defaults need to be coded at the places that the config is read. Also, many of Bodhi's settings actually do not define defaults and Bodhi will explode if some settings don't exist.
3. Some settings are used in more than once place (including the above cited example), introducing the risk that a setting will be interpreted differently in one place than it is in another (which could be especially bad if they choose different defaults).

There are a couple things we can do to help:

1. Have config.py interpret all the settings' types, so that the rest of the code can say things just like the above and get the right type (bool, int, string, whatever) without having to remember to cast it.
2. Put all defaults into config.py so that the bodhi config file can be mostly commented out. Some settings can't be commented out because they are pyramid settings, but the bodhi specific ones could all have sane defaults. We need to be careful not to change any existing defaults when doing this.

# Outside the scope

This is outside the scope of this ticket, but I wanted a place to write down some related thoughts.

Something else I've pondered is giving Bodhi a nicer config file format. AFAIK, we are actually stuck with the ini file for Pyramid settings, but there's no reason we couldn't split Bodhi-specific settings into a new config file. TOML or YAML would be nice. This way, the config file itself encodes the types of the settings rather than us doing awkward things like parsing a string that has commas and deciding that means a list.

### PR DESCRIPTION
**Pull request by bowlofeggs**
This pull request has two commits. The first is the primary goal of the PR, and gives Bodhi a central config validation system. This is a significant refactor, and a technical debt payment. I recommend reading its commit message for further details.

The first commit happened to raise test coverage to 85%, so the second commit requires test coverage to remain at 85%.

fixes #1489

### CODE REVIEW

**Comment by pypingou** [comment]
Looks like it is conflicting now :(

**Comment by bowlofeggs** [comment]
@pypingou I have resolved the conflict, and I have also altered the second commit to require 86% test coverage since @ryanlerch recently also merged another commit that increased test coverage to 85%. That means we've gotten it raised by 2% in a month!

**Comment by pypingou** [comment]
Pretty big one, but the bodhi_config file looks good

**Comment by bowlofeggs** [comment]
I have now rebased this PR on top of #1536, incorporating its new settings and adjusting it to work the new way that's introduced in this PR.

File: bodhi/server/mail.py [hunk 1]
```diff
@@ -182,7 +182,7 @@
 """,
         'fields': lambda agent, x: {
             'package': x.title,
-            'comment': x.comments[-1],
+            'comment': x.comments[-1].text,
```
Comment (ralphbean) [inline bodhi/server/mail.py]: This is a surprising change to find in this PR.

What's this about?

File: bodhi/server/models.py [hunk 2]
```diff
@@ -2208,8 +2206,7 @@ def check_karma_thresholds(self, db, agent):
         if self.autokarma and self._composite_karma[1] != 0:
             log.info("Disabling Auto Push since the update has received negative karma")
             self.autokarma = False
-            text = unicode(config.get(
-                'disable_automatic_push_to_stable', DEFAULT_DISABLE_AUTOPUSH_MESSAGE))
+            text = config.get('disable_automatic_push_to_stable', DEFAULT_DISABLE_AUTOPUSH_MESSAGE)
```
Comment (ralphbean) [inline bodhi/server/models.py]: Is this default still needed here?

File: setup.cfg [hunk 3]
```diff
@@ -2,7 +2,7 @@
 cover-erase=TRUE
 cover-html=TRUE
 cover-inclusive=TRUE
-cover-min-percentage=84
+cover-min-percentage=87
```
Comment (ralphbean) [inline setup.cfg:5]: Woohoo!!

**Comment by ralphbean** [comment]
In general, this change looks great.

I left two questions where I was genuinely uncertain about what should be done.

Basic :+1: from me to merge at your discretion.

File: bodhi/server/mail.py (hunk 1 above)
Comment (bowlofeggs) [reply bodhi/server/mail.py]: @ralphbean I had been having some encoding related issues when I was developing this patch early on, and I think I had done this to test that. I reverted this change and all tests pass so I think it's not needed. I'll back it out.

File: bodhi/server/models.py (hunk 2 above)
Comment (bowlofeggs) [reply bodhi/server/models.py]: @ralphbean It is not, good catch! I'll remove it too.

**Comment by bowlofeggs** [comment]
@ralphbean In response to your comments, I am applying this patch to my commit:

```
diff --git a/bodhi/server/mail.py b/bodhi/server/mail.py
index 6650014..8846400 100644
--- a/bodhi/server/mail.py
+++ b/bodhi/server/mail.py
@@ -182,7 +182,7 @@ To reply to this comment, please visit the URL at the bottom of this mail
 """,
         'fields': lambda agent, x: {
             'package': x.title,
-            'comment': x.comments[-1].text,
+            'comment': x.comments[-1],
             'updatestr': unicode(x)
         }
     },
diff --git a/bodhi/server/models.py b/bodhi/server/models.py
index 00a8039..6507686 100644
--- a/bodhi/server/models.py
+++ b/bodhi/server/models.py
@@ -47,11 +47,6 @@ from bodhi.server.util import (
 import bodhi.server.util
 
 
-DEFAULT_DISABLE_AUTOPUSH_MESSAGE = (
-    u'Bodhi is disabling automatic push to stable due to negative karma. '
-    u'The maintainer may push manually if they determine that the issue is not severe.')
-
-
 # http://techspot.zzzeek.org/2011/01/14/the-enum-recipe
 
 class EnumSymbol(object):
@@ -2206,7 +2201,7 @@ class Update(Base):
         if self.autokarma and self._composite_karma[1] != 0:
             log.info("Disabling Auto Push since the update has received negative karma")
             self.autokarma = False
-            text = config.get('disable_automatic_push_to_stable', DEFAULT_DISABLE_AUTOPUSH_MESSAGE)
+            text = config.get('disable_automatic_push_to_stable')
             self.comment(db, text, author=u'bodhi')
         elif self.stable_karma and self.karma >= self.stable_karma:
             if self.autokarma:
diff --git a/bodhi/tests/server/services/test_updates.py b/bodhi/tests/server/services/test_updates.py
index 95486c6..6994252 100644
--- a/bodhi/tests/server/services/test_updates.py
+++ b/bodhi/tests/server/services/test_updates.py
@@ -28,7 +28,7 @@ import mock
 from bodhi.server import main
 from bodhi.server.config import config
 from bodhi.server.models import (
-    BuildrootOverride, DEFAULT_DISABLE_AUTOPUSH_MESSAGE, Group, RpmPackage, Release,
+    BuildrootOverride, Group, RpmPackage, Release,
     ReleaseState, RpmBuild, Update, UpdateRequest, UpdateStatus, UpdateType, User)
 import bodhi.tests.server.functional.base
 
@@ -3246,8 +3246,7 @@ class TestUpdatesService(bodhi.tests.server.functional.base.BaseWSGICase):
         up.comment(self.db, u'Failed to work', author=u'ralph', karma=-1)
         up = self.db.query(Update).filter_by(title=resp.json['title']).one()
 
-        expected_comment = unicode(config.get(
-            'disable_automatic_push_to_stable', DEFAULT_DISABLE_AUTOPUSH_MESSAGE))
+        expected_comment = config.get('disable_automatic_push_to_stable')
         self.assertEquals(up.comments[2].text, expected_comment)
 
         up.comment(self.db, u'LGTM Now', author=u'ralph', karma=1)
```

I will rebase and merge this PR now - thanks for the review!

**Comment by ralphbean** [comment]
:unicorn:

**Pull request merged**